earth_enterprise/src/common/khxml/khxml.cpp

```diff
@@ -101,66 +125,195 @@ void setDefaultValues()
    initialDOMHeapAllocSize = 0x4000;
    maxDOMHeapAllocSize     = 0x20000;
    maxDOMSubAllocationSize = 0x1000;
+   terminateCache = false;
+   percent = 1.5;
+}
+
+void getInitValues()
+{
+    static bool callGuard = false;
+    if (callGuard) return;
+    callGuard = true;
+    setDefaultValues();
+    try
+    {
+        for(const auto& i : options)
+        {
+            config_parser.addOption(i);
+        }
+        config_parser.parse(XMLConfigFile.c_str());
+        config_parser.validateIntegerValues();
+        for (const auto& it : config_parser)
+        {
+            if (it.first == INIT_HEAP_SIZE)
+                initialDOMHeapAllocSize = std::stol(it.second);
+            else if (it.first == MAX_HEAP_SIZE)
+                maxDOMHeapAllocSize = std::stol(it.second);
+            else if (it.first == BLOCK_SIZE)
+                maxDOMSubAllocationSize = std::stol(it.second);
+            else if (it.first == PURGE)
+                terminateCache = std::stoi(it.second);
+            else
+            {
+                switch (std::stoi(it.second))
+                {
+                    case  1: percent = 0.00; break;
+                    case  2: percent = 0.75; break;
+                    case  3: percent = 1.50; break;
+                    case  4: percent = 2.25; break;
+                    case  5: percent = 3.00; break;
+                    default: percent = 1.50;
+                };
+            }
+        }
+    }
+    catch (const khConfigFileParserException& e)
+    {
+        notify(NFY_DEBUG, "%s , using default xerces init values", e.what());
+    }
+    try
+    {
+        validateXMLParameters();
+    }
+    catch (const XmlParamsException& e)
+    {
+        setDefaultValues();
+        notify(NFY_DEBUG, "%s, using default xerces init values", e.what());
+    }
 }
 
+void initXercesValues()
+{
+    getInitValues();
+    try
+    {
+        XMLPlatformUtils::Initialize(initialDOMHeapAllocSize,
+                                     maxDOMHeapAllocSize,
+                                     maxDOMSubAllocationSize);
+        notify(NFY_DEBUG, "XML initialization values: %s=%zu %s=%zu %s=%zu %s=%d %s=%f",
+               "initialDOMHeapAllocSize", initialDOMHeapAllocSize,
+               "maxDOMHeapAllocSize", maxDOMHeapAllocSize,
+               "maxDOMSubAllocationSize", maxDOMSubAllocationSize,
+               "purge cache", terminateCache,
+               "level", percent);
+    }
+    catch (const XMLException& toCatch)
+    {
+        notify(NFY_FATAL, "Unable to initialize Xerces: %s",
+               FromXMLStr(toCatch.getMessage()).c_str());
+    }
+}
+
+static khMutexBase xmlLibLock = KH_MUTEX_BASE_INITIALIZER;
+static khMutexBase checkTermLock = KH_MUTEX_BASE_INITIALIZER;
+static khMutexBase cacheLock = KH_MUTEX_BASE_INITIALIZER;
+static khMutexBase reinitLock = KH_MUTEX_BASE_INITIALIZER;
+
+class purgeGuard
```

Comment:
I think what you're trying to implement is a read/write lock, except that you have a write/purge lock. Writes can happen at the same time, but purges can't happen during writes or other purges. It would be safer if you used a standard read/write lock and did a write lock for purges and a read lock for writes (I know that's paradoxical, but it's the right choice in this situation).